### DISCUSSION TITLE
Close : CSV import tools: big numbers and limits.

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#192] [Ideas]
you who like big numbers, and limits :-)

2 files to import with the import csv too
[pb-import-1.csv](https://github.com/textbrowser/biblioteq/files/8357169/pb-import-1.csv)
[pb-import-2.csv](https://github.com/textbrowser/biblioteq/files/8357170/pb-import-2.csv)

I recognize that it is not common but an error in a script and the quantity column is crazy.
in my database there are 2 documents that had reached > 200 of quantity

**Comment by meteos77** [comment]
suggestion: 
in the GUI: bound the quantity widget because 999999999 seems to me a lot

long term suggestion :
Offer a button to pre-analyze the csv file to check some parameters.

for example
isbn13 starts with 978 or 979
the quantity is different from 1 ?
are the dates well formatted ?

even if it is not blocking, it would alert the user that in his file, some data are not common

**Reply by textbrowser** [reply]
This is an endless discussion. Suppose it's the import which guards what's coming into the database. OK. It cannot guard what happens after. Suppose each widget guards its field. It cannot guard what happens to the source code. Suppose the database guards the format of the data. The database can be modified.

In closing, the people are responsible for formatting their data. :)

**Reply by meteos77** [reply]
you may have noticed that I like to make suggestions :-)

**Reply by textbrowser** [reply]
You're ultimately responsible for your data.

**Reply by textbrowser** [reply]
Any tweaks to BiblioteQ which makes it impossible to use by others is a poor solution.

**Reply by meteos77** [reply]
in my mind BiblioteQ doesn't modify the data, it could only warn before the import (without modifying the data, it's up to the user to make the necessary modifications : we agree).

we have anyway the errors **after** the import.

advice : make a new database for each csv import and after checking the import make the merge of the 2 databases

**Reply by textbrowser** [reply]
Suppose you have 10,000 rows all of which have incorrect something (whatever that is). That's 10000 errors. When does it stop?

**Reply by textbrowser** [reply]
The problem with data purifiers is that they are specific. They add zero value to the tool itself. I intended for the software to be data agnostic. Then people asked for more and more to make it data aware. It didn't improve the software. It made it more difficult to change. It decreased its flexibility.

**Reply by textbrowser** [reply]
LibreOffice has a spell checker. It doesn't produce errors automatically. It allows you to enter whatever you want. The spell checker is a separate tool.

**Reply by textbrowser** [reply]
The data are format-less.

**Reply by meteos77** [reply]
the analyzer informs at the 1st doubt encountered and asks the user to Stop or Continue

if I take an analogy on my idea it would be close to a 
(as in connector technology: you can't plug the plug in the wrong way)

my translator : polarizer 
but rather https://en.wikipedia.org/wiki/Poka-yoke

Libreoffice integrates well the corrector inside

**Reply by meteos77** [reply]
https://en.wikipedia.org/wiki/Poka-yoke
https://fr.wikipedia.org/wiki/D%C3%A9trompeur#Poka-yoke_et_baka-yoke

a paragraph in the French wiki that is not in the English wiki
maybe a cultural difference?

Associated thought system

The interest of a polarizer is to avoid a faulty production.

The nuance may not seem very interesting compared to the initial definition: in fact, it is the basis of the whole Japanese system of thinking in production (see the example of Toyota), in which "doing quality" is prioritized over "controlling quality". In a positive way, the deception becomes a device to allow production only if it is good.

For many years, the importance of the cultural approach to quality problems has been overlooked. This limitation has notably led to the failure of quality circles (launched by Kaoru Ishikawa in Japan in 1962) in France, where the Japanese language is still considered "very foreign" [unclear].

**Reply by textbrowser** [reply]
So we're back to the beginning. I can spend a limitless amount of time perfecting a small part of the software, namely, the importer. Say, 10 continuous weeks of perfecting data, producing reports, notifications, etc.

Then, someone modifies the data in SQLite using a separate tool and complains that the software is producing an incorrect output.

Conclusion? I spent 10 weeks of life for nothing.

**Reply by textbrowser** [reply]
This is not balance but extremism. :P Balance is some responsibility with the customer and some responsibility with the software. Time is not a vast quantity where perfection is above all else.

**Reply by textbrowser** [reply]
> the analyzer informs at the 1st doubt encountered and asks the user to Stop or Continue
> 
> if I take an analogy on my idea it would be close to a (as in connector technology: you can't plug the plug in the wrong way)
> 
> my translator : polarizer but rather https://en.wikipedia.org/wiki/Poka-yoke
> 
> Libreoffice integrates well the corrector inside

This approach is basic and assumes many things without stating them. For example, it may be beneficial for closed and controlled processes. In open or irregular processes, the process is totally flawed. Why? Let's assume that BQ purifies the data-entry process. That is, data entered through BQ is guaranteed to be correct within some well-defined guidelines. Once a separate entry is applied, BQ is asleep. It cannot guarantee anything beyond its limited boundaries. The effort for an upfront barrier is totally lost once BQ is circumnavigated. BQ cannot guarantee that another tool is not used, the database is replaced and/or modified, the source is changed. It is an irregular process and a regular solution is itself irregular.

**Reply by meteos77** [reply]
**first of all respect and total thanks for the time spent on free software development**
my proposal is a mountain while I thought small hill -> you are not a magician?

if user is out of BiblioteQ; it's out of subject; I don't think my suggestions are out of BQ
example : with a sledgehammer I broke my screen and the title column does not display well in BQ :-) can you do something ?

**Reply by meteos77** [reply]
agree balance : interests / time spent / time saved / quality
not being a programmer, the notion "time spent for development" is vague

**Reply by meteos77** [reply]
I don't agree, the concept of unavoidable human errors and processes that eliminate defects at the source seems good to me

[see car reliability](https://fr.wikipedia.org/wiki/Classements_de_marques_de_voitures_par_indice_de_fiabilit%C3%A9) (in the lead ... Japanese cars ... this may not be a coincidence)

**Reply by meteos77** [reply]
by the way I see in BiblioteQ a certain form and on certain fields and I think it is wonderful
your concept of "Database Enumeration" is perfect, it presents the freedom for the user to choose the data and restricts this choice in the articles
as counter examples :
too much freedom : being able to enter any format in the user's email field doesn't seem appropriate to me (a simple check *@*.*)
too much restriction : the state field for the members, if out of US ; no right to put the info ?

for me external tools are irrelevant ; if the user doesn't want to use BiblioteQ ... too bad for him :-)

**Reply by textbrowser** [reply]
The software should inspire other developers. It's not intended as the ultimate solution. In 50,000 years, it'll be irrelevant. :) More like 100. Even 50. 20.

**Reply by meteos77** [reply]
and yes, it may disappear in a more or less near future like us :-)

but at the moment, it is still quite complete thanks to you and finding developers who are interested in a free library management software with all the necessary notions is very difficult.

**Reply by textbrowser** [reply]
Niche software is ignored because developers want fame.

**Reply by textbrowser** [reply]
Like music, independent video games, strange literature, art, etc.

**Reply by meteos77** [reply]
I have studied the different existing softwares and have been following your development for only 5 months and according to certain criteria : BiblioteQ is the only one to offer certain functionalities
- multiOS
- sqlite & sql language
- c++ language for development
...

unfortunately for our library, it still misses some essential functionalities for our use but with time I think it will be one of the most complete and one of the best (in the world of the free and the owner)
with a TODOLIST at zero it is the number 1 in its category

**Reply by meteos77** [reply]
yet it's easier to get the glory when the competition is minimal :-)

**Reply by textbrowser** [reply]
And remember, most of its completion status is not because of actual requests. It's because of? Imagination.

https://user-images.githubusercontent.com/10701156/160432682-93574e82-0c1b-43f7-a464-a7ff3cf9f587.mp4

**Reply by meteos77** [reply]
[good imagination french version](https://www.youtube.com/watch?v=gqtgiibmLdU)

**Reply by meteos77** [reply]
perhaps that brings a more in the software and has the use one feels it well (I have evil with the other software and to make the same work of importation and test)

**Reply by meteos77** [reply]
the only program at the same level of quality as BiblioteQ that I have encountered is tellico but it is for collection management and not for library management.
also developed in qt & c++ but no sql ; no loan management, no user management) according to my test of November 2021

long life to both projects anyway!

**Reply by textbrowser** [reply]
Glory suppresses creativity, interoperability (mine is best paradigm), imagination. When these are suppressed, innovation is too.

**Comment by textbrowser** [comment]
https://github.com/textbrowser/biblioteq/commit/925fe5962e0290afecf20f4d07df9129195b0535
